src/lib/core/ReferenceCounted.h

```diff
@@ -80,4 +81,46 @@ class ReferenceCounted
     count_type mRefCount = kInitRefCount;
 };
 
+/// An variant of ReferenceCounted, with atomic counter.
+template <class Subclass, class Deletor = DeleteDeletor<Subclass>, int kInitRefCount = 1>
+class AtomicReferenceCounted
+{
+public:
+    using count_type = uint32_t;
+
+    AtomicReferenceCounted() : mRefCount(kInitRefCount) {}
+
+    /** Adds one to the usage count of this class */
+    Subclass * Retain()
+    {
+        if (mRefCount == std::numeric_limits<count_type>::max())
+        {
+            abort();
+        }
+        ++mRefCount;
```

Comment:
The read and increment all need to be done within a single atomic operation for this to be thread safe.  Else, two threads can see the ref count is 0xFFFFFFFE and then two threads can increment and wrap to 0.

```suggestion
        if (mRefCount++ == std::numeric_limits<count_type>::max())
        {
            abort();
        }
```

Reply:
> The TCP endpoint pool is not linked previously because there is no TCP usage.

I can address this part separately, as I'm abstracting the EndPoint implementations (for which I'm just waiting on this PR, since the constructors need to run).

Reply:
Looks like @kpschoedel's comment is not related to my suggestion change.

But atomics have been resolved by removing them.

Reply:
Removed atomic